Giphy Block Registration

jeffersonrabb opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

Registers the Giphy block and handles Server-Side Rendering. SSR approach is being used because IFRAME element will be stripped out of post content on WP.com. Discussion of this here: p5TWut-gm-p2. 

Companion `wp-calypso` PR: https://github.com/Automattic/wp-calypso/pull/29638

#### Changes proposed in this Pull Request:
- Registers the Giphy block 
- Callback to render the Giphy block. 

#### Testing instructions:

- Connect Jetpack
- Navigate to Settings->Jetpack Constants, check `JETPACK_BETA_BLOCKS`, and Save.

1) Add the block
2) Go to Giphy, find any Gif you like, copy the URL and paste into the block. The gif should show up.
3) Paste in a Giphy URL following this syntax: http://i.giphy.com/4ZFekt94LMhNK.gif. Past the URL into the block. The gif should show up.
4) Type any text in to the search field. DIfferent Gifs should show up for anything you search for. 
5) Verify that the text field appears when you click on the block. 
6) Verify that the text field disappears after 3.5 seconds of inactivity, unless no search text has been provided in which case it remains visible.
7) Verify caption behaves much like the core Image block.
8) Verify block alignment options all work as expected.


Complete testing instructions found here: https://github.com/Automattic/wp-calypso/pull/29638

#### Changelog

* Block editor: addition of a new Giphy block.

jeffersonrabb committed:
Register Giphy block.

jeffersonrabb committed:
Giphy block, server-side rendering to protect the iFrame from being stripped out.

jeffersonrabb committed:
Change to top padding handling. Attribute name change to 'paddingTop', inclusion of '%' in the saved attribute.

sirreal commented on modules/blocks.php:
```diff
@@ -60,6 +66,35 @@ function jetpack_tiled_gallery_load_block_assets( $attr, $content ) {
+			<iframe src='<?php echo( esc_attr( $giphy_url ) ); ?>' title='<?php echo( esc_attr( $search_text ) ); ?>'></iframe>
+		</figure>
+		<?php if ( $caption ) : ?>
+			<figcaption class="caption"><?php echo( esc_html( $caption ) ); ?></figcaption>
```
🤔 This is a concern. We can't `esc_html` here because it's RichText in the editor. We _expect_ HTML.

Should we use [`wp_kses`](https://developer.wordpress.org/reference/functions/wp_kses/)?

Also, `figcaption` must be first or last child of `figure` [according to MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/figcaption). We'll need to rework the layout a bit.

Are we planning to keep captions, or can we forget this question altogether?

sirreal replied:
Tracking the captions or no question: https://github.com/Automattic/wp-calypso/issues/30345

sirreal replied:
We've decided to keep captions.

I used [`wp_kses_post`](https://developer.wordpress.org/reference/functions/wp_kses_post/) to escape this in c412741.

sirreal committed:
Clean up quotes and parens

- Remove parens around echo
- Add missing style quotes
- Use double quotes consistently

sirreal commented:
I pushed a commit to clean up the HTML a bit. There were some missing as well as mixed single and double quotes.

I removed the parens around the echo arguments, they're not necessary and they seem to be mostly omitted in this code base.

jeffersonrabb committed:
Change caption to P from FIGCAPTION, add new class.

sirreal committed:
Use wp_kses_post to escape caption HTML

sirreal commented on modules/blocks.php:
```diff
@@ -60,6 +66,37 @@ function jetpack_tiled_gallery_load_block_assets( $attr, $content ) {
+function jetpack_giphy_block_load_assets( $attr ) {
+	$align       = isset( $attr['align'] ) ? $attr['align'] : 'center';
+	$style       = 'padding-top:' . $attr['paddingTop'];
+	$giphy_url   = isset( $attr['giphyUrl'] ) ? $attr['giphyUrl'] : '//giphy.com/embed/ZgTR3UQ9XAWDvqy9jv';
```
Shouldn't we abort rendering anything if we don't have a URL? It seems unlikely users will insert the block without searching and want the placeholder gif.

jeffersonrabb replied:
Not sure I would agree. Seems like the view-side should always be a representation of what the user sees in the editor, and if they drag in the block but perform no search, they'll see the placeholder. Hopefully in this case, seeing the placeholder in the Preview is a loud reminder that they should use the block to search for the content they want, but I think that suppressing the block entirely ends up being more confusing overall.

sirreal replied:
> Seems like the view-side should always be a representation of what the user sees in the editor, and if they drag in the block but perform no search, they'll see the placeholder.

I agree with this premise. Which leads me to believe that we should use a true Gutenberg placeholder as the placeholder in the editor.

I wasn't involved in that part of the discussion, but I would like to raise the concern here before moving on. We can change this pretty simply in a follow-up.

Should we show the editor "placeholder" gif on the frontend if the user hasn't changed the gif? Should we use a standard Gutenberg placeholder?

cc: @thomasguillot

jeffersonrabb replied:
Resolved in https://github.com/Automattic/wp-calypso/pull/30402 and https://github.com/Automattic/jetpack/pull/10989/commits/bd0a2818a78154c0ef7e71e30cc7fbd15d91f43c

sirreal commented on modules/blocks.php:
```diff
@@ -60,6 +66,37 @@ function jetpack_tiled_gallery_load_block_assets( $attr, $content ) {
+			<iframe src="<?php echo esc_attr( $giphy_url ); ?>" title="<?php echo esc_attr( $search_text ); ?>"></iframe>
+		</figure>
+		<?php if ( $caption ) : ?>
+			<p class="wp-block-jetpack-giphy-caption"><?php echo wp_kses_post( $caption ); ?></p>
```
Highlighting for reviewers the use of `wp_kses_post` here.

sirreal reviewed (dismissed):
This looks good and tests well. I have [a lingering doubt](https://github.com/Automattic/jetpack/pull/10989/files#r250289212) but it can be addressed in a follow-up PR.

jeherve commented on modules/blocks.php:
```diff
@@ -60,6 +66,37 @@ function jetpack_tiled_gallery_load_block_assets( $attr, $content ) {
+	?>
+	<div class="wp-block-jetpack-giphy align<?php echo esc_attr( $align ); ?>">
+		<figure style="<?php echo esc_attr( $style ); ?>">
+			<iframe src="<?php echo esc_attr( $giphy_url ); ?>" title="<?php echo esc_attr( $search_text ); ?>"></iframe>
```
Should we use `esc_url` instead of `esc_attr` here for the URL?

jeffersonrabb replied:
Yup, resolved here: https://github.com/Automattic/jetpack/pull/10989/commits/71d7d458d4aac2c58b2446de4495bac5a94faa03

jeherve reviewed (dismissed):
I only have minor remarks about this PR (see below), but I do have some feedback about the block itself.

*****

Should Giphy be capitalized in the block description?

![image](https://user-images.githubusercontent.com/426388/51679695-e57f3d00-1fdf-11e9-966c-61d1b6d85161.png)

For some reason the additional classes don't get added to the wrapper:

![image](https://user-images.githubusercontent.com/426388/51679780-211a0700-1fe0-11e9-9fc8-125f56b8a112.png)

I was a bit surprised by the block icon; I would have expected it to be an image or a Giphy logo instead of this:

![image](https://user-images.githubusercontent.com/426388/51680099-2297ff00-1fe1-11e9-8389-76afbb12f907.png)

jeffersonrabb commented:
> I was a bit surprised by the block icon; I would have expected it to be an image or a Giphy logo instead of this:

cc: @thomasguillot

jeffersonrabb commented:
> I only have minor remarks about this PR (see below), but I do have some feedback about the block itself.

@jeherve since these all relate to `wp-calypso` would you mind filing as issues and I'll address them immediately?

thomasguillot commented:
> I was a bit surprised by the block icon; I would have expected it to be an image or a Giphy logo instead of this:

@jeherve This is Giphy's logo (flat version).

One thing I previously mentioned would be:

We rename the block to something more generic that users would actually know: "Gif". And then we replace the Giphy icon by the Image icon

jeherve commented:
> This is Giphy's logo (flat version).

Ha, my bad! :)

jeffersonrabb committed:
Rename Giphy block to Gif.

jeffersonrabb committed:
Support for Additional CSS Class in Gif block. Fixes https://github.com/Automattic/wp-calypso/issues/30385

jeffersonrabb committed:
Do not render block if Giphy URL is unset. This approach replaces the default Giphy URL.

jeffersonrabb commented:
> I have [a lingering doubt](https://github.com/Automattic/jetpack/pull/10989/files#r250289212) but it can be addressed in a follow-up PR.

This "lingering doubt" is the placeholder state, correct? If so, it's addressed in: https://github.com/Automattic/wp-calypso/pull/30402 and https://github.com/Automattic/jetpack/pull/10989/commits/bd0a2818a78154c0ef7e71e30cc7fbd15d91f43c

jeffersonrabb committed:
Use esc_url() to sanitize Giphy URL.

jeffersonrabb committed:
Check for existence of top padding attribute before use.

jeffersonrabb commented on modules/blocks.php:
```diff
@@ -60,6 +66,37 @@ function jetpack_tiled_gallery_load_block_assets( $attr, $content ) {
+ */
+function jetpack_giphy_block_load_assets( $attr ) {
+	$align       = isset( $attr['align'] ) ? $attr['align'] : 'center';
+	$style       = 'padding-top:' . $attr['paddingTop'];
```
Done in 
https://github.com/Automattic/jetpack/pull/10989/commits/27491b47c8789fb6383b7c078d280635a05a8d0a

jeffersonrabb committed:
Renamed function to better describe use.

jeffersonrabb commented on modules/blocks.php:
```diff
@@ -3,9 +3,15 @@
+jetpack_register_block(
+	'giphy',
+	array(
+		'render_callback' => 'jetpack_giphy_block_load_assets',
```
Done in https://github.com/Automattic/jetpack/pull/10989/commits/091c70928c851ae0825334adfe16c6f3bd31a25c

jeffersonrabb commented on modules/blocks.php:
```diff
@@ -60,6 +66,37 @@ function jetpack_tiled_gallery_load_block_assets( $attr, $content ) {
+ *
+ * @return string
+ */
+function jetpack_giphy_block_load_assets( $attr ) {
```
Done https://github.com/Automattic/jetpack/pull/10989/commits/091c70928c851ae0825334adfe16c6f3bd31a25c

sirreal commented on modules/blocks.php:
```diff
@@ -60,6 +66,50 @@ function jetpack_tiled_gallery_load_block_assets( $attr, $content ) {
+ */
+function jetpack_gif_block_render( $attr ) {
+	$align       = isset( $attr['align'] ) ? $attr['align'] : 'center';
+	$padding_top = isset( $attr['paddingTop'] ) ? $attr['paddingTop'] : 0;
```
This would indicate we don't know the image size and the element would be either:

- be invisible because it would be inside a 0 height container
- overlap the following content because of its absolute positioning

is that correct? I haven't confirmed which is the case, but either seems to lead to a broken experience. Should we abort rendering the block if this is missing or `0` (falsy)?

sirreal replied:
This is not correct. A fallback size is provided and the iframe is resized to fit within the container.

sirreal commented on modules/blocks.php:
```diff
@@ -60,6 +66,50 @@ function jetpack_tiled_gallery_load_block_assets( $attr, $content ) {
+	$search_text = isset( $attr['searchText'] ) ? $attr['searchText'] : '';
+	$caption     = isset( $attr['caption'] ) ? $attr['caption'] : null;
+
+	if ( ! $giphy_url ) {
```
It seems that we _require_ both of these to correctly render the block:

```suggestion
	if ( ! $giphy_url || ! $padding_top ) {
```

sirreal reviewed (commented):
I left one suggestion.

sirreal committed:
Abort if paddingTop is missing

sirreal committed:
Revert "Abort if paddingTop is missing"

This reverts commit 94392ce5cc509ccc8d917a12931311e42b7337fd.

sirreal reviewed (approved):
I think this is in a good place and the serialized block structure has likely reached some stability. I think this is ready to land any time.

jeherve reviewed (approved):
I like it. Let's ship it!

https://giphy.com/gifs/ship-wood-launch-ACqTaHVTEvEZO

spacedmonkey commented:
For anyone interested, I have open a ticket on core and on the gutenberg repo. 

https://github.com/WordPress/gutenberg/issues/13426
https://core.trac.wordpress.org/ticket/46067
